### PR TITLE
devicetree: Add a zephyr/zephyr.dts debugging aid that has the final devicetree

### LINKED ISSUE #22272: aggregated devicetree source file needs to be restored to build directory
#21785 included what became 8d8b06978c577bff24006aa3b3c875d9dae820d0 which removes the generated aggregate devicetree source file.

This file is critical for diagnosing problems in devicetree content and syntax.  That it's not used by the tooling anymore does not mean it's not used.  There is no other reasonable way to confirm that specific overlays are being made available to the generation scripts.

We do seem to have `board.dts.pre.d` and `board.dts.pre.tmp` still available, but there's no promise that they have meaningful content, unlike `.dts_compile` which I could rely on to be an accurate representation of the devicetree content used when generating code.

### PR DESCRIPTION
**Pull request by ulfalizer**
(Screenshots still show `merged.dts`, but I renamed it to `zephyr.dts`.)

Adds a new `zephyr/zephyr.dts` output that holds the final devicetree, as a debugging aid. `zephyr.dts` preserves labels in phandle and path references too.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22272.

Also redraws a diagram in the devicetree section of the manual to be accurate after the old devicetree scripts were removed.

Before:
io-before

After:
io-after

`zephyr.dts` is included in the build overview diagram now too:

Skärmbild från 2020-01-29 10-26-43

Commits:

```
dts: edtlib: Add an EDT.dts_source attribute that holds the source code

EDT.dts_source is the source code of the loaded .dts file, after merging
nodes.
```

```
dts: gen_defines.py: Save final devicetree to zephyr/zephyr.dts

Add a '--dts-out <file>' flag to gen_defines.py that saves the final
devicetree to <file>, in DTS format, using the new EDT.dts_source
attribute.

Handy to have available as a debugging aid. Unused otherwise.
```

```
doc: devicetree: Update outdated diagram and mention zephyr.dts

zephyr_dt_inputs_outputs.svg still shows the output from dtc being used,
but dtc is unused (except for finding warnings/errors) after the old
devicetree scripts were removed in commit c8c35f76ab ("scripts: dts:
Remove deprecated extract_dts_includes.py script").

Update zephyr_dt_inputs_outputs.svg to show how things are done now.
Also include the new zephyr.dts debugging aid in it. Tweak the
formatting a bit too.

Add zephyr.dts to the diagram in the build overview section too, and
mention zephyr.dts in the text of the devicetree and build overview
pages.

Remove zephyr_dt_inputs_outputs.png and use zephyr_dt_inputs_outputs.svg
directly. Many other places the documentation include SVGs directly, and
there haven't been any complaints, so it probably works fine. The .png
and .svg versions had also drifted out of sync.

Piggyback a link from the devicetree page to the build overview page, to
make it easier to discover.

(I used draw.io to update the diagrams.)
```

### CODE REVIEW

**Comment by ulfalizer** [comment]
Skipped the board name in `zephyr/merged.dts` since the board is already known. Nice to have a simple, consistent filename.

**Review by pabigot — CHANGES_REQUESTED** [review]
I could bikeshed the name (final.dts, application.dts, image.dts) but other than not really liking merged.dts I can't immediately identify a clearly superior solution.  Except maybe zephyr.dts for consistency with zephyr.elf and zephyr.map and other final outputs that end up in that directory.

Yeah, I'd like zephyr.dts.

Are we guaranteed that the devicetree source produced by this library can be processed by dtc?  Because I can see doing that.

**Comment by ulfalizer** [comment]
> I could bikeshed the name (final.dts, application.dts, image.dts) but other than not really liking merged.dts I can't immediately identify a clearly superior solution. Except maybe zephyr.dts for consistency with zephyr.elf and zephyr.map and other final outputs that end up in that directory.
> 
> Yeah, I'd like zephyr.dts.
> 

`merged.dts` seems more specific to me, so I like it better.

I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.

So `merged.dts` might be better from a psychology perspective.

> Are we guaranteed that the devicetree source produced by this library can be processed by dtc? Because I can see doing that.

Should be. Otherwise there's a bug, and it should be pretty stable, because it gets tested a lot indirectly in `testdtlib.py`.

**Comment by ulfalizer** [comment]
`final.dts` doesn't seem too bad either though. Not against going with that (except I'd have to wrestle draw.io for a few minutes again).

**Comment by pabigot** [comment]
> `merged.dts` seems more specific to me, so I like it better.
> 
> I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.

I don't think that makes sense from a developer perspective.  We tend to look for `zephyr.{bin,elf,hex,lst,map,stat}` in that directory as being associated with the final image product.  `merged.dts` doesn't suggest it's associated with the final product; it could be an intermediate file.  If I want a dt blob that corresponds to the image needs for some reason, I'd look for `zephyr.dts` to produce it.

@mbolivar ?

**Comment by ulfalizer** [comment]
> > `merged.dts` seems more specific to me, so I like it better.
> > I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.
> 
> I don't think that makes sense from a developer perspective. We tend to look for `zephyr.{bin,elf,hex,lst,map,stat}` in that directory as being associated with the final image product. `merged.dts` doesn't suggest it's associated with the final product; it could be an intermediate file. If I want a dt blob that corresponds to the image needs for some reason, I'd look for `zephyr.dts` to produce it.

Ah, yeah, looking at it that way it does make some sense. Still would've been confused though, because I never thought if it that way (anything with zephyr in the name = final output).

Then I won't fight it at least. I'll rename it if it helps the PR go in faster.

**Comment by mbolivar** [comment]
I think I'd prefer either not renaming it at all (BOARD.dts_compiled) or using zephyr.dts.

Actually, could you create BOARD.dts_compiled for zephyr 2.2 in addition to zephyr.dts, even if its contents just say "go look at zephyr.dts instead", similarly to how you've kept the old header files around for now?

IMO developers will be looking for this file, and will be confused by its disappearance. Having a shim for one release will help.

**Comment by ulfalizer** [comment]
Renamed the file to `zephyr.dts`.

> Actually, could you create BOARD.dts_compiled for zephyr 2.2 in addition to zephyr.dts, even if its contents just say "go look at zephyr.dts instead", similarly to how you've kept the old header files around for now?
> 
> IMO developers will be looking for this file, and will be confused by its disappearance. Having a shim for one release will help.

Added. `cmake/dts.cmake` now writes a dummy `<BOARD>.dts_compiled`.

File: cmake/dts.cmake [hunk 1]
```diff
@@ -209,6 +210,11 @@ if(SUPPORTS_DTS)
     message(FATAL_ERROR "new extractor failed with return code: ${ret}")
   endif()
 
+  # A file that used to be generated by 'dtc'. zephyr.dts is the new
+  # equivalent.
+  file(WRITE ${PROJECT_BINARY_DIR}/${BOARD}.dts_compiled
+       "See zephyr.dts for the final merged devicetree.")
+
```
Comment (SebastianBoe) [inline cmake/dts.cmake]: Let's add a comment about removing this after the next Zephyr release.

We don't want this noise here forever.

Comment (ulfalizer) [reply cmake/dts.cmake]: Added "will be removed in Zephyr 2.3."

File: cmake/dts.cmake [hunk 2]
```diff
@@ -198,6 +198,7 @@ if(SUPPORTS_DTS)
   --bindings-dirs ${DTS_ROOT_BINDINGS}
   --conf-out ${DEVICETREE_CONF}
   --header-out ${DEVICETREE_UNFIXED_H}
+  --dts-out ${PROJECT_BINARY_DIR}/merged.dts # As a debugging aid
```
Comment (pabigot) [inline cmake/dts.cmake]: This is still generating `merged.dts` rather than `zephyr.dts`.

**Review by pabigot — CHANGES_REQUESTED** [review]

**Comment by galak** [comment]
Generally LGTM, once you guys decide on the details.

File: cmake/dts.cmake (hunk 2 above)
Comment (ulfalizer) [reply cmake/dts.cmake]: Oops... search fail. Fixed.

Had missed a spot in the docs too.

**Review by pabigot — APPROVED** [review]

**Review by SebastianBoe — APPROVED** [review]

**Comment by mbolivar** [comment]
From the docs commit message:

> Many other places the documentation include SVGs directly, and there haven't been any complaints, so it probably works fine.

FWIW, I created a PNG specifically because the SVG was not rendering properly in Firefox on Linux. Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it. I left the SVG in there to make future edits easier.

Have you tested this on a few different browsers?

**Comment by mbolivar** [comment]
> Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it.

Found the comment with the details: https://github.com/zephyrproject-rtos/zephyr/pull/19630#issuecomment-538582774

**Comment by ulfalizer** [comment]
> FWIW, I created a PNG specifically because the SVG was not rendering properly in Firefox on Linux. Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it. I left the SVG in there to make future edits easier.
> 
> Have you tested this on a few different browsers?

That one seems to render fine in Firefox at least:

Skärmbild från 2020-01-29 17-12-58

Looks good in Chromium too.

Wonder if something might be wonky with dimensions if other ones are broken.

**Review by mbolivar — APPROVED** [review]
I double-checked the documentation in my browser and the SVG looks good there too. Looks good to me, thanks!

**Review by galak — APPROVED** [review]

**Pull request merged**